cdap-tms/src/main/java/io/cdap/cdap/messaging/store/leveldb/LevelDBMessageTable.java

```diff
@@ -64,83 +73,102 @@ byte getType() {
     }
   }
 
-  private final DB levelDB;
-  private final TopicMetadata topicMetadata;
+  private final LevelDBPartitionManager partitionManager;
 
-  LevelDBMessageTable(DB levelDB, TopicMetadata topicMetadata) {
-    this.levelDB = levelDB;
-    this.topicMetadata = topicMetadata;
+  LevelDBMessageTable(LevelDBPartitionManager partitionManager) {
+    this.partitionManager = partitionManager;
   }
 
-  private void checkTopic(TopicId topicId, int generation) {
-    Preconditions.checkArgument(this.topicMetadata.getTopicId().equals(topicId), "Not allowed to use table with a " +
-      "different topic id. Table's topic Id: {}. Specified topic id: {}", this.topicMetadata.getTopicId(), topicId);
-    Preconditions.checkArgument(this.topicMetadata.getGeneration() == generation, "Not allowed to use table with " +
-                                  "a different generation id. Table's generation: {}. Specified generation: {}",
-                                this.topicMetadata.getGeneration(), generation);
+  @Override
+  protected CloseableIterator<RawMessageTableEntry> scan(ScanRequest scanRequest) throws IOException {
+    Collection<LevelDBPartition> partitions = partitionManager.getPartitions(scanRequest.getStartTime());
+    if (partitions.isEmpty()) {
+      return CloseableIterator.empty();
+    }
+    RawMessageTableEntry tableEntry = new RawMessageTableEntry();
+    BiFunction<byte[], byte[], RawMessageTableEntry> decodeFunction = (key, value) -> {
+      Map<String, byte[]> columns = decodeValue(value);
+      return tableEntry.set(MessageTableKey.fromBytes(key), columns.get(TX_COL), columns.get(PAYLOAD_COL));
+    };
+
+    return new PartitionedDBScanIterator<>(partitions.iterator(), scanRequest.getStartRow(), scanRequest.getStopRow(),
+                                           decodeFunction);
   }
 
   @Override
-  protected CloseableIterator<RawMessageTableEntry> read(byte[] startRow, byte[] stopRow) {
-    final DBScanIterator iterator = new DBScanIterator(levelDB, startRow, stopRow);
-    final RawMessageTableEntry tableEntry = new RawMessageTableEntry();
-    return new AbstractCloseableIterator<RawMessageTableEntry>() {
-      private boolean closed = false;
-
-      @Override
-      protected RawMessageTableEntry computeNext() {
-        if (closed || (!iterator.hasNext())) {
-          return endOfData();
-        }
+  protected void persist(Iterator<RawMessageTableEntry> entries) throws IOException {
+    Map<Long, WriteBatch> writeBatches = new TreeMap<>();
```

Comment:
Entries coming from the Iterator are actually sorted by publish time already. So instead of using a tree map, you can just loop and roll to the next LevelDBPartition if the partition changed (you may need a method in the partitionManager to have something like `LevelDBPartition partitionManager.getOrCreate(time, @Nullable currentPartition)`. This can keep the linear performance for the persist call.

Reply:
This could also simplify the error handling code below.

Reply:
ah good to know